### PR TITLE
ThreadLocal based Inspektr PrincipalResolver

### PR DESCRIPTION
**Pull request by dima767**
This PR introduces a ThreadLocal based storage for current Authentication object as well as Inspektr PrincipalResolver based on that state. This is in response to already complex and brittle current Joinpoint-based PrincipalResolver that does complex Joinpoint parsing and is very tightly coupled to CAS API which in the latest iterations has changed significantly and therefore this resolver is not able to resolve proper principal at all the configured audit points.

This PR also has the side effect that this ThreadLocal holder for current authentication could be used by other interested downstream components without tight coupling to all the internal CAS components just to get a currently in-progress authentication/principal, should they need it.

### CODE REVIEW

File: cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import static java.util.stream.Collectors.joining;
+
+/**
+ * ThreadLocal based holder for current set of credentials and/or authentication object for any current
+ * CAS authentication transaction. Useful for making this information available to all the interested CAS
+ * components that are not tightly coupled with core CAS APIs, for example audit principal resolver component, etc.
+ * <p>
+ * The thread local state carried by this class should be set by core CAS components processing core authentication and
+ * CAS protocol events e.g. <code>AbstractAuthenticationManager</code>, <code>CentralAuthenticationServiceImpl</code>, etc.
```
Comment (mmoayyed) [inline cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java]: Minor suggestion: try using {@link}. Helps with future refactoring, if the class should be renamed, etc.

File: cas-server-webapp-config/src/main/java/org/apereo/cas/config/CasFiltersConfiguration.java [hunk 2]
```diff
@@ -99,4 +100,18 @@ public FilterRegistrationBean requestParameterSecurityFilter() {
...
         return bean;
     }
+
+
+    /**
+     * Current credentials and authentication clearing filter.
+     *
+     * @return the current credentials and authentication clearing filter
+     */
+    public FilterRegistrationBean currentCredentialsAndAuthenticationClearingFilter() {
```
Comment (mmoayyed) [inline cas-server-webapp-config/src/main/java/org/apereo/cas/config/CasFiltersConfiguration.java]: No @Bean here?

File: cas-server-webapp/build.gradle [hunk 3]
```diff
@@ -35,7 +35,7 @@ bootRun {
     }
 
     addResources = true
-    jvmArgs = ["-Xrunjdwp:transport=dt_socket,address=5000,server=y,suspend=n"]
+    //jvmArgs = ["-Xrunjdwp:transport=dt_socket,address=5000,server=y,suspend=n"]
```
Comment (mmoayyed) [inline cas-server-webapp/build.gradle]: Revert this back?

**Comment by mmoayyed** [comment]
Thanks @dima767 Looks excellent. Few minor questions here and there.

File: cas-server-webapp/build.gradle (hunk 3 above)
Comment (mmoayyed) [inline cas-server-webapp/build.gradle]: Never mind. Dont need this anymore.

File: cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java (hunk 1 above)
Comment (dima767) [inline cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java]: Will add the link, but that will require compile dependency on the module in question :-)

File: cas-server-webapp-config/src/main/java/org/apereo/cas/config/CasFiltersConfiguration.java (hunk 2 above)
Comment (dima767) [inline cas-server-webapp-config/src/main/java/org/apereo/cas/config/CasFiltersConfiguration.java]: Oops... Will add :-)

File: cas-server-webapp/build.gradle (hunk 3 above)
Comment (dima767) [inline cas-server-webapp/build.gradle]: Yeah, this messed up my IDEA debug config, as I run CAS app within IDEA. This is really not necessary. I'll remove it altogether.

File: cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java (hunk 1 above)
Comment (mmoayyed) [inline cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/CurrentCredentialsAndAuthentication.java]: Gotcha. Lets keep it then.

**Pull request merged**